### ISSUE TITLE
Type def component in router config

### ISSUE THREAD

**Issue by blephy** [gregberge/loadable-components#637]
Hi,

I can't get rid of TypeScript except by typing `any` at one place. I need help on the following code :

```ts
/** My imports */
import loadable from '@loadable/component'
import pMinDelay from 'p-min-delay'
import React from 'react'

import Loader from '@/components/loader'

export const Home = loadable(() => pMinDelay(import(/* webpackChunkName: 'home' */ './home'), 200), {
  fallback: <Loader />
})
export const NotFound = loadable(() => pMinDelay(import(/* webpackChunkName: 'notFound' */ './not-found'), 200), {
  fallback: <Loader />
})
```ts
/** My App component, which is in fact the main router
import '@/styles/index.scss'

import React, { ReactNode } from 'react'
import { Helmet } from 'react-helmet'
import { BrowserRouter as Router, Redirect, Route, Switch } from 'react-router-dom'

import ErrorBoundary from '@/components/error-boundary'

import { IRoute } from './app.routes'

export interface IProperties {
  routes: IRoute[]
}

export default class App extends React.PureComponent<IProperties> {
  static defaultProps = {
    routes: [{}]
  }

  render(): ReactNode {
    const { routes } = this.props

    return (
      <>
        <ErrorBoundary>
          <Helmet defaultTitle={process.env.APP_TITLE} titleTemplate={`%s | ${process.env.APP_TITLE}`}>
            <link rel='canonical' href={process.env.SERVER_BASE_URL} />
          </Helmet>
          <Router>
            <Switch>
              {routes.map(({ component: Component, ...rest }) => (
                <Route
                  exact={rest.exact || false}
                  path={rest.path}
                  key={`${rest.name}_${rest.path}`}
                  render={routeProperties => <Component {...routeProperties} />}
                />
              ))}
              <Redirect to='/404' />
            </Switch>
          </Router>
        </ErrorBoundary>
      </>
    )
  }
}
```

```ts
/** My route config, here is the probleme */
import { LoadableComponent } from '@loadable/component'

import { Home, NotFound } from '@/views'

export interface IRoute {
  name: string
  exact?: boolean
  path: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: LoadableComponent<any>
}

const routes: IRoute[] = [
  {
    name: 'home',
    exact: true,
    path: '/',
    component: Home
  },
  {
    name: 'notFound',
    path: '/404',
    component: NotFound
  }
]

export default routes
```

How can i properly type the `component: LoadableComponent<any>` ?

When i try with `unknown` i'm getting this error on `<Component />` inside my App component : 

```txt
Type '{ history: History<unknown>; location: Location<unknown>; match: match<any>; staticContext?: StaticContext; }' has no properties in common with type 'IntrinsicAttributes & { fallback?: Element; } & { children?: ReactNode; }'.
```

**Comment by ThisIsMissEm** [comment]
This is because of the missing the types on `loadable()`, you need to do: 

```
loadable<RouteComponentProps<Params, Context, State>>(() => /* the import stuff */)
```

Then you can also just do:

```
<Route
  exact={rest.exact || false}
  path={rest.path}
  key={`${rest.name}_${rest.path}`}
  component={Component}
/>
```

**Comment by ThisIsMissEm** [comment]
@gregberge would you be in favour of adding to the documentation usage of loadable component with react-router and typescript?

I finally worked out I needed to do as the above states, but it took me hours to work in out correctly, and everyone seemed to solve this incorrectly by using a `component` prop as a function, which causes needless re-rendered each time the route changes.

**Comment by theKashey** [comment]
> This is because of the missing types on loadable(), you need to do: ....

No, that is not the right way to "fix" the problem. Let's refer to [typescript definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/loadable__component/index.d.ts#L71) for `loadable`

```ts
declare function loadableFunc<Props>(
    loadFn: (props: Props) => Promise<DefaultComponent<Props>>,
    options?: OptionsWithoutResolver<Props>,
): LoadableComponent<Props>;
```

What you did - you have overriden `Props`
How it shall work - `Props` shall be inferred from `DefaultComponent`. But they were not. And that is the root cause.

In this particular example probaly  `component: LoadableComponent<any>` is the problem.
It shall be as per @ThisIsMissEm - `component: React.ComponentType<RouteComponentProps>`

**Comment by ThisIsMissEm** [comment]
okay, took me a little to understand your comment @theKashey, it sounded like you were either explaining my own commemt to me or telling me I was wrong. I'd actually missed the `IRoute` definition when commenting, but even if you used:

```component: React.ComponentType<RouteComponentProps>``` 

It wouldn't work. `component` Is explicitly a LoadableComponent, in this code, so just:

```
component: LoadableComponent<RouteComponentProps>
``` 

Should be enough, I think

**Comment by theKashey** [comment]
Lets refer to what is `LoadableComponent` - it is just a component with props, plus `fallback`, plus static methods like `preload`
```ts
export type LoadableComponent<Props> = React.ComponentType<Props & { fallback?: JSX.Element }> &
    LoadableComponentMethods<Props>;
```

However `Router` should don't care. For the Router it's just `Component`, and the question who else need to know that it is actually Loadable.
For the `preload` use case it's better to use `type guards`
```ts
const isPreloadableComponent=<T extends {}>(component: ComponentType<T>):component is Loadable<T> => 'preload' in component;
```

> explaining my own commemt to me or telling me I was wrong.

Almost every time you have to specify generic, like you did in your example, is because something is preventing a "real" value to be inferred, which is an indication of a problem.

👉 In other words `loadable<PleaseNeverOverrideMe>`

**Comment by ThisIsMissEm** [comment]
@theKashey please, I promise you that I've done my work here. Try doing: 

```
const mycomponent = loadable(() => import('./my-component'))

<Route path="/" component={mycomponent}>
```

I promise you that won't work. React Router requires a component with `RouteComponentProps` as props.

**Comment by theKashey** [comment]
> React Router requires a component with RouteComponentProps as props.

Well, not quite, here is it's defenition
`component?: React.ComponentType<RouteComponentProps<any>> | React.ComponentType<any>;`

As a bigger example:
```ts
const X = loadable(() => import(`./Page`))
<Route component={X}/> // no error
<Route
        path="otherPath"
        render={routeProperties =>
        // @ts-expect-error 
          <X {...routeProperties} /> // expect error as long..
        }
/>
```
> 👉 Error:(20, 16) TS2559: Type '{ history: History<unknown>; location: Location<unknown>; match: match<any>; staticContext?: StaticContext; }' has no properties in common with type 'IntrinsicAttributes & { fallback?: Element; } & { children?: ReactNode; }'.

These properties really does not exists on component, and TypeScript informing you about it.

----

Why `component` is working? Because it's not _requires a component with RouteComponentProps as props._ - TypeScript uses structutal types (DuckTyping) and requires component to be __less__ than `RouteComponentProps`, to a subset of it, not "greater".
In other words - you can miss some properly, even all, but not add something `Route` is not able to provide to you.

__And `component` accepts `any` as well__.

----

I've also creates [typescript example with react-router](https://github.com/gregberge/loadable-components/blob/master/examples/typescript/src/client/App.tsx) to reproduce this issue, and the short term goal it to compile-and-run those examples during CI.

And specifically the problem of this issue __is expected__. It is "correct" to have it -> https://github.com/gregberge/loadable-components/blob/master/examples/typescript/src/client/App.tsx#L24

**Comment by blephy** [comment]
This is the problem @theKashey . You point it well. I'm using `<Route ... render=` and not `<Route ... component=` because i want to be able to pass props in the component, if he needs to include a router too (child router ...).

But anyway, didn't find the right way

**Comment by theKashey** [comment]
Or you can asses params you need in some other way, like via `useRouterStuff`

**Comment by blephy** [comment]
I did manage to get it working with `LoadableComponent<RouteComponentProps>`

I don't know if this is a TS version or something, but i promess this was a pain some weeks ago ! :)

Ty @theKashey and @ThisIsMissEm
